hugr-core/src/hugr/persistent.rs

```diff
@@ -0,0 +1,387 @@
+pub type PersistentReplacement = SimpleReplacement<PatchNode>;
+
+/// A patch that can be applied to a [`PersistentHugr`] or a
+/// [`CommitStateSpace`] as an atomic commit.
+///
+/// Commits are cheap to clone: they are reference-counted pointers to the
+/// patch data. They also maintain strong references to the ancestor commits
+/// that the patch may depend on (i.e. other patches that must be applied
+/// before `self` can be applied).
+///
+/// Currently, patches must be [`SimpleReplacement`]s.
+#[derive(Debug, Clone, From)]
+#[repr(transparent)]
+pub struct Commit(RelRc<CommitData, ()>);
+
+impl Commit {
+    /// Create a commit from a simple replacement.
+    ///
+    /// Requires a reference to the commit state space that the nodes in
+    /// `replacement` refer to.
+    ///
+    /// The replacement must act on a non-empty subgraph, otherwise this
+    /// function will return an [`InvalidCommit::EmptyReplacement`] error.
+    ///
+    /// If any of the parents of the replacement are not in the commit state
+    /// space, this function will return an [`InvalidCommit::UnknownParent`]
+    /// error.
+    pub fn try_from_replacement(
+        replacement: PersistentReplacement,
+        graph: &CommitStateSpace,
+    ) -> Result<Commit, InvalidCommit> {
+        if replacement.subgraph().nodes().is_empty() {
+            return Err(InvalidCommit::EmptyReplacement);
+        }
+        let parent_ids = replacement.invalidation_set().map(|n| n.0).unique();
+        let parents = parent_ids
+            .map(|id| {
+                if graph.contains_id(id) {
+                    Ok(graph.get_commit(id).clone())
+                } else {
+                    Err(InvalidCommit::UnknownParent(id))
+                }
+            })
+            .collect::<Result<Vec<_>, _>>()?;
+        let rc = RelRc::with_parents(
+            replacement.into(),
+            parents.into_iter().map(|p| (p.into(), ())),
+        );
+        Ok(Self(rc))
+    }
+
+    fn as_relrc(&self) -> &RelRc<CommitData, ()> {
+        &self.0
+    }
+
+    /// Get the set of nodes removed by `self`.
+    fn deleted_nodes(&self) -> impl Iterator<Item = PatchNode> + '_ {
+        let replacement = match self.0.value() {
+            CommitData::Base(_) => None,
+            CommitData::Replacement(replacement) => Some(replacement),
+        };
+        replacement
+            .into_iter()
+            .flat_map(|r| r.subgraph().nodes())
+            .copied()
+    }
+
+    delegate! {
+        to self.0 {
+            fn value(&self) -> &CommitData;
+        }
+    }
+}
+
+impl From<Commit> for RelRc<CommitData, ()> {
+    fn from(Commit(data): Commit) -> Self {
+        data
+    }
+}
+
+impl<'a> From<&'a RelRc<CommitData, ()>> for &'a Commit {
+    fn from(rc: &'a RelRc<CommitData, ()>) -> Self {
+        // SAFETY: Commit is a transparent wrapper around RelRc
+        unsafe { mem::transmute(rc) }
+    }
+}
+
+/// A Hugr-like object that supports persistent mutation.
```

Comment:
I think "supports persistent mutation" is a rather odd way to describe it: maybe "... that includes its mutation history"? I'm still a bit doubtful about the name, though I don't have a great alternative. Would something like `HugrRecord` or `HugrHistoryGraph` be workable? They sound less "Hugr-like" though. Hmm ... perhaps `PersistentHugr` is best after all.

Reply:
Agreed, "persistent mutation" is an awkward phrasing. I like and have adopted your suggested phrasing.

That being said, I do like the `PersistentHugr` name. I think using the adjective "persistent" is correct and helpful in this context, referring to an established pattern of functional programming. I'm open to other suggestions though.